### ISSUE TITLE
Youtube subtitle does not speed up

### ISSUE THREAD

**Issue by 29988122** [igrigorik/videospeed#537]
Description:
On youtube, the video speed controller adjustment, and the youtube built-in speed adjustment function are not synced.

That is, when adjusting speed via youtube built-in funcion, video speed controller stays at 1.0x, and the opposite is the same.

Issue:
Youtube subtitle only speeds up when adjusting via built-in speed control, not via video speed controller - i.e., when adjust speed via video speed controller, the subtitle will de-sync with the video.

Possible workaround:
Put a dirty fix to make both option sync, or make video speed controller control the subtitle speed of youtube. (someties you need more than 2x, in lots of the country people don't listen, they watch subtitle even if the content it's their mother tongue)

**Comment by ChadBailey** [comment]
I don't understand what you're trying to report. The speed adjustment works on YouTube, and the subtitles go at the adjusted speed on my machine. Are you saying that the speed of the subtitles are going at 1x while the playback rate is 2x?

**Comment by 29988122** [comment]
>Are you saying that the speed of the subtitles are going at 1x while the playback rate is 2x?
Only when the speed is modified by Video Speed Controller.
Sorry for the confusion caused.

Situation 1:
Youtube speed 1x, Video Speed Controller 2x or more
subtitle desynced

Situation 2:
Youtube speed 1.25x ~ 2x, Video Speed Controller speed settings irrelevant
subtitle synced

**Comment by Eisys** [comment]
> Situation 1:
> Youtube speed 1x, Video Speed Controller 2x or more
> subtitle desynced

Yeah this happens to me too. I don't use CC that much but when I do, I have to reset VSC to 1x, then choose speed from the YT player menu. Then the CC are in sync.
Otherwise, if VSC speed is 2x for example, the YT player menu still shows the speed as "normal" so it tries to show the CC at a normal speed too.

**Comment by ChadBailey** [comment]
Interesting, I'll test more to see if I can reproduce it. It's certainly possible I hadn't watched for long enough for the two to diverge

**Comment by jacobcolbert** [comment]
You have any specific examples? I just watched a few videos to conclusion at 4x speed and they were fine.

**Comment by Eisys** [comment]
>https://youtu.be/-tXFD6Hwk3k

Yeah that's exactly what happens. The player tries to keep up but the CC isn't perfectly in sync.
It doesn't go completely out of sync more and more as the video plays but it's not really in sync either. Atleast not as well as if the player itself controlled/knew the speed.

**Comment by ChadBailey** [comment]
Alright, I was able to reproduce - it seems I was correct in my assumption that I had not watched long enough for the subtitles to lag noticeably behind.

I do have some concerns about this request.

The first is, why wouldn't you just use the built in speed adjustments since they work as intended? You can access them from the GUI, or just use `<` to slow down and `>` to speed up (be sure to hold shift).

The second concern is, if we did tightly couple the youtube player speed with playback rate, it would very likely mean we would end up being forcefully limited to a maximum of 2x speed on youtube, which would be unacceptable for me personally - that's the entire reason I got this plugin in the first place was to go beyond 2x speed playback rate.

**Comment by 29988122** [comment]
>that's the entire reason I got this plugin in the first place was to go beyond 2x speed playback rate.

This is my intent, too. (to go beyond 2x)

If, somehow you can make video speed controller > 2x, but, in the same time, still keep youtube sub in sync, that'd be the best.

However, I understand that:
1. This "fix" would be out of scope, too much work, youtube change its player so our bandaid fix broke, etc. Hence wontfix. Completely reasonable. (If I want this feature I should study extention coding and make a PR myself, not forcing it onto volunteering maintainers.)
2. It's impossible due to extention permission thing.

In that situation, I'd suggest that, maybe, we can put an extra setting option, such as "keep youtube subtitle in sync", and make a description text which says that this makes the speed only adjustable from 0.25x to 2x.

People love this addon, and I myself only explicitly use this addon's keyboard shortcut to adjust speed - that's why I encountered this issue in the first place. 
Putting an extra option/reminder might seem unnecessary, but considering the popularity of youtube, I'd personally try implementing such an option in exchange for elegance.

Just my two cents anyway, you're the boss. 

And again, thanks for help maintaining such a useful addon. Literally lifesaving when watching Netflix and many, many video websites. : p

**Comment by ChadBailey** [comment]
Your report is a good report, and let's be clear, I'm no boss! lol

Irrespective of if anything can be done about it, you are making a great contribution yourself by making this report so that others can come and read for themselves if they are experiencing the same problem.

**Comment by jacobcolbert** [comment]
Thanks for the reports. I try to take accessibility concerns very seriously so I would definitely like to see this fixed. As the single biggest source of online video I think youtube would warrant some domain specific code within reason. This is definitely a change on their part as I know the internal speed tracker used to keep pace with vsc initiated speed changes. I will take a look and see if I can figure out a good strategy but welcome assistance.

**Comment by jacobcolbert** [comment]
Okay. So some progress on this. It is very easy to access the YT api from the top context on youtube.com. Setting the speed like this synchronizes both 

```
player=document.getElementById('movie_player')
player.setPlaybackRate(2)
```

A few comments:

1. Accessing the top context requires injecting code directly into the page. This seems like an officially sanctioned technique, but I'm not sure if there is anything we should be wary about here. It will also require communicating with the injected code via dispatching events.
2. This should also work on YT embeds since there is a VSC instance running in the iframe of the video.
3. It will round speeds to the nearest YT accessible speed. Most importantly it won't go past 2x speed.

Despite these limitations and requirement of a chunk of domain specific code I personally think it is worth pursuing. It is necessary for Deaf and foreign language speakers, and useful for everyone.

I'm imagining a checkbox in the option menu asking if you want to sync YT captions noting that it will restrict you to 2x speed on YT. Ideally it would then only use the message passing if this choice is selected and captions were turned on. There remarkably doesn't seem to be a way to tell if they are turned on using the API, but a hacky way is to look at the state of the captions button.

**Comment by ChadBailey** [comment]
Since the functionality is already available, why wouldn't a person needing this functionality disable VSC for youtube and use the built-in speed controller?

**Comment by jacobcolbert** [comment]
True. I still would like for an integrated solution. Is Video Speed
Controller really a Video Speed Controller if we tell a large share of the
population to turn it off on the behemoth of internet video?

I agree it is not a top priority since it is get aroundable, but I would
like to see what an attempt looks like. I will work on one by the new year.

**Comment by ChadBailey** [comment]
@jacobcolbert I wasn't asking a rhetorical question there, I really am having trouble understanding the design rationale.

If I'm understanding your proposal right, you're suggesting that we tightly couple VSC speed with YouTube's speed on the YT website. The value of doing this is it will speed up subtitles along with video playback rate. Did I get that right?

If I did get it right, then what I'm saying is all of those features exist in today's world with no changes necessary by using YouTube's built in controls... why wouldn't users of subtitles just use YT's speed controls? Seems like an easy and obvious solution... I must be missing something here?

I can think of some other options to satisfy the concern - _if_ they are technically feasible.

1. *Find a way to speed up the subtitles to match
2. Provide a warning on the configuration page that subtitles aren't in sync on YouTube when using VCS
3. *Automatically disable VSC when a user has subtitles enabled
4. *Bind a warning toast to the routine that enables subtitles to warn the user that they must use YouTube's speed controller for subtitles

* Might not be possible without an upstream change in the Chrome extension API to allow us access to event listeners

This is based on my very limited domain knowledge in this area, so I may be missing something obvious

**Comment by ChadBailey** [comment]
> Interesting, I'll test more to see if I can reproduce it. It's certainly possible I hadn't watched for long enough for the two to diverge

I actually was correct here.... I did more testing and found that on some videos the closed captions will completely keep up speed. I think there are 2 things impacting this.

First, when you enable closed captioning it seems that this is what sets the "rate"... try hitting "c" twice to toggle them off/on and see if that helps with sync issues

Second, YT seems to have 2 types of closed captions - one auto-generated and another supplied by the video maker. The auto-generated ones actually seem to stay in sync far better than the type supplied by the video maker. To see a demonstration of this, look at video https://www.youtube.com/watch?v=BHnMItX2hEQ which has auto-generated subtitles.

**Comment by ChadBailey** [comment]
Should this issue be closed? I'm not certain we ever came to consensus on this one

**Comment by 29988122** [comment]
I'd still want a solution, if possible for you guys - whether it's an ugly one(bandaid only for youtube, prone to break) or an elegant one. I guess this is also what people need since as youtube is the largest video site online.

@ChadBailey The sync issue is fine on my side for a few months - but it's back now, out of sync without manualy adjusting the built-in controller(as you've said, I tried both an auto-generated one and a manually add one. The former is better, but out of sync still exists). Click "c" for a few times doesn't work.

It would be greatly appreciated if any fix can be applied - it's out of my reach when I tried reading the codebase of videospeedcontroller. Unfortunate that I'm not good enough to contribute myself.

**Comment by ChadBailey** [comment]
I've tagged as needs investigation so I can dig deeper into this again. I need to tend to more pressing bugs first then I'll take a deeper look.

Meanwhile, can I have you give me a sampleset of at least 4 videos (the more the better) of greater than 10 minutes each (the longer the better, for testing purposes) and include preferably the speeds used to reproduce? Please try to include videos that it does happen and videos that it does not happen on.

**Comment by 29988122** [comment]
Here is one of the many examples:
https://www.youtube.com/watch?v=BT-oBrbrSqo
the video in the video(lol): https://www.youtube.com/watch?v=gxAaO2rsdIs

I've demostrated both built-in function and videospeedcontroller function, and compared the two on subtitles.

This can be reliably reproduced for almost every video on youtube.

**Comment by ChadBailey** [comment]
Thanks for taking the effort to record a video explaining! Ironically I've already seen this video, love 3blue1brown's videos!

I plan to dig deep into the open issues this weekend so I will take a closer look when I do (hate saying a date because, ya know, stuff happens lol).

**Comment by VimalMollyn** [comment]
Hey! Here's a link to an example video, where the captions lag behind (I tried it at 2.8x) - https://www.youtube.com/watch?v=ZA4JkHKZM50

I'm on Firefox 80.0.1 (64-bit) for Ubuntu. I'm noticing a combination of skipped captions and(or) lag. By skipped, I mean that some sentences' captions are skipped (presumably because of the >2x speed).

Thanks!

**Comment by 29988122** [comment]
I wonder is this will ever be implemented

Still hope for the best (I don't know shit about front-end development)

Thanks!

**Comment by Noobeley** [comment]
Has this been fixed yet? I love watching slow YouTube videos at 3x speed but that means I have to turn subtitles off and that's an awful experience. 

I think it IS possible to trick youtube into thinking you're at 3x speed. Ever hear of Youtube vanced? It's a custom client for youtube android and it allows you to change the playback speed beyond 2x, AND the subtitles sync perfectly. Can you look at their code and implement that?

**Comment by danielzgtg** [comment]
Vanced can modify the subtitle code, while we don't modify YouTube's code here.

I have an idea: How about we just programmatically click the toggle captions button (or press c per others above), but at a rate of two times at the start of every second. Then add CSS to remove the "English auto generated" overlay. This could work because the subtitles are re-synced every time the button is toggled.

**Comment by Noobeley** [comment]
> Vanced can modify the subtitle code, while we don't modify YouTube's code here.
> 
> I have an idea: How about we just programmatically click the toggle captions button (or press c per others above), but at a rate of two times at the start of every second. Then add CSS to remove the "English auto generated" overlay. This could work because the subtitles are re-synced every time the button is toggled.

That's a great idea. Could you program it in? You could make it so subtitles stick on screen and are only "toggled on+off" in the backend.

**Comment by Noobeley** [comment]
I managed to write a script that toggles the subtitles on and off rapidly using AI but i have no clue how to make it so the subtitles stick so it's not flashing so much I can't see

**Comment by danielzgtg** [comment]
@Noobeley Sorry, I just found out that toggling causes a flickering because YouTube always shows any changes on the next frame.

After poking around testing the effects of changing a bunch of different functions, I found that modifying the `delay` passed to `setTimeout` works. The full code is at #1185.
